### PR TITLE
Nix solidity package: call solc just one and extract all the files af…

### PR DESCRIPTION
**Pull request by gbalabasquer**
…ter if needed

### CODE REVIEW

File: nix/solidity-package.sh [hunk 1]
```diff
@@ -25,7 +19,26 @@ if [ "$doCheck" == 1 ]; then
...
+    ((count++))
+    name="${path#*:}"
+    contract=$(echo "$data" | jq '.[''"'"$path"'"'']')
+    echo "$contract" | jq '.["abi"]' -r > "$DAPP_OUT/$name.abi"
+    echo "$contract" | jq '.["bin"]' -r > "$DAPP_OUT/$name.bin"
+    echo "$contract" | jq '.["bin-runtime"]' -r > "$DAPP_OUT/$name.bin-runtime"
+    echo "$contract" | jq '.["metadata"]' -r > "$DAPP_OUT/$name.metadata"
+  done
+  echo
+  info "warning: --extract is slow and may have filename collisions. All build data can be found in $DAPP_JSON.\\n"
```
Comment (icetan) [inline nix/solidity-package.sh:38]:
```suggestion
```

File: nix/solidity-package.sh [hunk 2]
```diff
@@ -25,7 +19,26 @@ if [ "$doCheck" == 1 ]; then
...
-cp -r $src $out/dapp/$name/src
-cp -r lib $out/dapp/$name/lib
-cp -r out $out/dapp/$name/out
+if [ "$extract" == 1 ]; then
+  mapfile -t contracts < <(<"$DAPP_JSON" jq '.contracts|keys[]' -r | sort -u -t: -k2 | sort)
+  data=$(<"$DAPP_JSON" jq '.contracts' -r)
+  count=1
+  total=${#contracts[@]}
+  for path in "${contracts[@]}"; do
+    info -ne "Extracting build data... [$count/$total]\\r"
```
Comment (icetan) [inline nix/solidity-package.sh:28]: Don't think `info` is a defined function here.

```suggestion
```

File: nix/solidity-package.sh [hunk 3]
```diff
@@ -1,20 +1,14 @@
...
-  dir=${dir#$DAPP_SRC}
-  dir=${dir#/}
-  mkdir -p "$DAPP_OUT/$dir"
-  (set -x; solc --overwrite $REMAPPINGS --allow-paths $DAPP_SRC $solcFlags --abi --bin --bin-runtime -o "$DAPP_OUT/$dir" "$x")
-  json_file=$DAPP_OUT/$dir/${x##*/}.json
-  (set -x; solc $REMAPPINGS --allow-paths $DAPP_SRC $solcFlags $jsonopts "$x" >"$json_file")
-done
+mkdir -p "$DAPP_OUT"
+mapfile -t files < <(find "${DAPP_SRC}" -name '*.sol')
+(set -x; solc "$REMAPPINGS" "$jsonopts" "$solcFlags" /=/ "${files[@]}" > "${DAPP_JSON}")
```
Comment (icetan) [inline nix/solidity-package.sh:11]: Where is `DAPP_JSON` declared?

File: nix/solidity-package.sh [hunk 4]
```diff
@@ -25,7 +19,26 @@ if [ "$doCheck" == 1 ]; then
...
+  mapfile -t contracts < <(<"$DAPP_JSON" jq '.contracts|keys[]' -r | sort -u -t: -k2 | sort)
+  data=$(<"$DAPP_JSON" jq '.contracts' -r)
+  count=1
+  total=${#contracts[@]}
+  for path in "${contracts[@]}"; do
+    info -ne "Extracting build data... [$count/$total]\\r"
+    ((count++))
+    name="${path#*:}"
+    contract=$(echo "$data" | jq '.[''"'"$path"'"'']')
+    echo "$contract" | jq '.["abi"]' -r > "$DAPP_OUT/$name.abi"
```
Comment (icetan) [inline nix/solidity-package.sh:32]: Need to add a subdirectory with the name of the dapp project.

File: nix/solidity-package.sh [hunk 5]
```diff
@@ -25,7 +19,26 @@ if [ "$doCheck" == 1 ]; then
...
-cp -r out $out/dapp/$name/out
+if [ "$extract" == 1 ]; then
+  mapfile -t contracts < <(<"$DAPP_JSON" jq '.contracts|keys[]' -r | sort -u -t: -k2 | sort)
+  data=$(<"$DAPP_JSON" jq '.contracts' -r)
+  count=1
+  total=${#contracts[@]}
+  for path in "${contracts[@]}"; do
+    info -ne "Extracting build data... [$count/$total]\\r"
+    ((count++))
+    name="${path#*:}"
```
Comment (icetan) [inline nix/solidity-package.sh:30]: This `name` might be overwriting `name` set by the derivation name.

File: nix/solidity-package.sh (hunk 3 above)
Comment (gbalabasquer) [reply nix/solidity-package.sh:11]: Right, I'm hardcoding it to `dapp.sol.json` now. If that is not an option as generates conflicts for different packages we will define something different.

File: nix/solidity-package.sh (hunk 5 above)
Comment (gbalabasquer) [reply nix/solidity-package.sh:30]: Changing it just in case.

File: nix/solidity-package.sh (hunk 2 above)
Comment (gbalabasquer) [reply nix/solidity-package.sh:28]: Yes you are right, I'm simplifying it to just one echo.

**Comment by gbalabasquer** [comment]
Closing this PR as it was actually made by mistake. We will create a new one as soon as this is finished.

**Pull request closed without merging**